generator/sbpg/targets/resources/c/new/sbp_messages_template.h

```diff
@@ -41,114 +42,405 @@
+   * @return Size of (((comment_name))) in wire representation
+   */
+  uint8_t (((prefix)))_packed_len(const (((string_type))) *s);
+
+  /**
+   * Query (((comment_name))) for remaining space
+   *
+   * Returns the number of bytes (not including NULL terminator) which can be added to (((comment_name))) before it exceeds the maximum size of the field in wire representation
+   *
+   * @param msg (((msg_type))) instance
+   * @return Maximum number of bytes that can be appended to the existing string
+   */
+  uint8_t (((prefix)))_space_remaining(const (((string_type))) *s);
 
   ((*- if f.encoding == "unterminated" or f.encoding == "null_terminated" *))
-  bool (((m.name|convert_unpacked)))_(((f.name)))_set(sbp_(((f.encoding)))_string_t *s, const char *new_str);
-  bool (((m.name|convert_unpacked)))_(((f.name)))_printf(sbp_(((f.encoding)))_string_t *s, const char *fmt, ...) SBP_ATTR_FORMAT(2,3);
-  bool (((m.name|convert_unpacked)))_(((f.name)))_vprintf(sbp_(((f.encoding)))_string_t *s, const char *fmt, va_list ap);
-  bool (((m.name|convert_unpacked)))_(((f.name)))_append_printf(sbp_(((f.encoding)))_string_t *s, const char *fmt, ...) SBP_ATTR_FORMAT(2,3);
-  bool (((m.name|convert_unpacked)))_(((f.name)))_append_vprintf(sbp_(((f.encoding)))_string_t *s, const char *fmt, va_list ap);
-  const char *(((m.name|convert_unpacked)))_(((f.name)))_get(const sbp_(((f.encoding)))_string_t *s);
+  /**
+   * Set (((comment_name)))
+   *
+   * Erase any existing content and replace with the specified string
+   *
+   * This function will return true if the new string was successfully applied. If the specified string is longer than can be represented in wire encoding this function will return false
+   *
+   * @param msg (((msg_type))) instance
+   * @param new_str New string
+   * @return true on success, false otherwise
+   */
+  bool (((prefix)))_set( (((-string_type))) *s, const char *new_str);
+
+  /**
+   * Set (((comment_name))) with printf style formatting
+   *
+   * Erase any existing content and replace with the formatted string
+   *
+   * This function will return true if the new string was successfully applied. If the operation would end up overflowing the maximum size of this field in wire encoding the existing contents will be erased and this function will return false.
+   *
+   * @param msg (((msg_type))) instance
+   * @param fmt printf style format string
+   * @return true on success, false otherwise
+   */
+  bool (((prefix)))_printf( (((-string_type))) *s, const char *fmt, ...) SBP_ATTR_FORMAT(2,3);
+
+  /**
+   * Set (((comment_name))) with printf style formatting
+   *
+   * Identical to #(((prefix)))_printf except it takes a va_list argument
+   *
+   * @param msg (((msg_type))) instance
+   * @param fmt printf style format string
+   * @param ap Argument list
+   * @return true on success, false otherwise
+   */
+  bool (((prefix)))_vprintf( (((-string_type))) *s, const char *fmt, va_list ap);
```

Comment:
the more I think about it, are these function really needed? if they are of a specific string type, wouldn't users just call the generic function for the string type?

another question would be, do we need to specify the string as the first argument or should we just pass in the struct pointer for the msg type given that the function name is already quite verbose?

Reply:
I've written it like this so that users of libsbp don't need to be aware of the maximum encoded length of a string field.

The implementation for these functions has a generated, const max length which gets passed in to the generic string encoder/decoder functions. It is purposefully kept separate from an instance of a message struct because it is intrinsic to the type and shouldn't ever be customised by the libsbp user.

The printf family of functions ended up being very useful in later work when building up certain messages like log. It will also allow removing the use of std::string in the settings registry in pvt-runner.